### PR TITLE
feat: turn snapshot prune async

### PR DESCRIPTION
**Pull request by blindchaser**

### CODE REVIEW

File: sei-db/sc/memiavl/db.go [hunk 1]
```diff
@@ -88,6 +89,9 @@ type DB struct {
 	mtx sync.Mutex
 	// worker goroutine IdleTimeout = 5s
 	snapshotWriterPool *pond.WorkerPool
+
+	// closed indicates the DB is shutting down
+	closed uint32
```
Comment (yzang2019) [inline sei-db/sc/memiavl/db.go]: Can we rename this to pruningInProgress?

File: sei-db/state_db/sc/flatkv/catchup.go [hunk 2]
```diff
@@ -0,0 +1,511 @@
+package flatkv
```
Comment (yzang2019) [inline sei-db/state_db/sc/flatkv/catchup.go]: Don't think we need this file

File: sei-db/sc/memiavl/db.go [hunk 3]
```diff
@@ -532,6 +542,9 @@ func (db *DB) pruneSnapshots() {
 		db.logger.Error("failed to find first snapshot", "err", err)
 	}
 
+	if db.streamHandler == nil {
+		return
+	}
 	if err := db.streamHandler.TruncateBefore(utils.VersionToIndex(earliestVersion+1, db.initialVersion.Load())); err != nil {
```
Comment (yzang2019) [inline sei-db/sc/memiavl/db.go:547]: Can we add a log here for how long the whole pruning process take?

File: sei-db/sc/memiavl/db.go [hunk 4]
```diff
@@ -491,9 +495,15 @@ func (db *DB) checkBackgroundSnapshotRewrite() error {
 
 // pruneSnapshot prune the old snapshots
 func (db *DB) pruneSnapshots() {
+	if atomic.LoadUint32(&db.pruningInProgress) == 1 {
```
Comment (yzang2019) [inline sei-db/sc/memiavl/db.go]: Maybe use compareAndSet?

File: sei-db/sc/memiavl/db.go [hunk 5]
```diff
@@ -532,6 +541,9 @@ func (db *DB) pruneSnapshots() {
 		db.logger.Error("failed to find first snapshot", "err", err)
 	}
 
+	if db.streamHandler == nil {
+		return
```
Comment (Kbhat1) [inline sei-db/sc/memiavl/db.go:545]: should we log something here?

File: sei-db/sc/memiavl/db.go [hunk 6]
```diff
@@ -491,9 +493,16 @@ func (db *DB) checkBackgroundSnapshotRewrite() error {
 
 // pruneSnapshot prune the old snapshots
 func (db *DB) pruneSnapshots() {
-	// wait until last prune finish
-	db.pruneSnapshotLock.Lock()
-	defer db.pruneSnapshotLock.Unlock()
+	// CAS: only one prune can run at a time
+	if !db.pruningInProgress.CompareAndSwap(false, true) {
```
Comment (masih) [inline sei-db/sc/memiavl/db.go:497]: What if a prune takes just long enough to overlap with the next prune? 
In that case the new prune cycle will be skipped and we end up with 2 cycles of unpruned state in the next prune

**Review by masih — COMMENTED** [review]
As is this is fine to test the theories discussed in the sync.But for a deeper fix i would like to understand the edge cases this change rises.

File: sei-db/sc/memiavl/db.go (hunk 6 above)
Comment (blindchaser) [reply sei-db/sc/memiavl/db.go:497]: yup, looking into the code. we will need to optimize some pruning logic and where/when pruning happens. creating a ticket for that.

Comment (yzang2019) [reply sei-db/sc/memiavl/db.go:497]: Each prune should prune all snapshots that need to be pruned, not just delete one snapshot. We also should make the pruning logic to be triggered after each commit, instead of only after snapshot creation
